sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala

```diff
@@ -471,11 +471,30 @@ class InMemoryCatalog(
       table: String,
       partialSpec: Option[TablePartitionSpec] = None): Seq[CatalogTablePartition] = synchronized {
     requireTableExists(db, table)
-    if (partialSpec.nonEmpty) {
-      throw new UnsupportedOperationException(
-        "listPartition with partial partition spec is not implemented")
+
+    partialSpec match {
+      case None => catalog(db).tables(table).partitions.values.toSeq
+      case Some(partial) =>
+        if (partial.size == getTable(db, table).partitionColumnNames.length) {
+          val matchedPartition = catalog(db).tables(table).partitions.get(partial)
+          Seq(matchedPartition.getOrElse {
+            val specString = partial.map { case (k, v) => s"$k=$v" }.mkString("{", ", ", "}")
+            throw new AnalysisException("Partition not found: " + specString)
+          })
+        } else {
+          catalog(db).tables(table).partitions.toSeq.collect {
+            case (spec, partition) if isPartialPartitionSpec(partial, spec) => partition
+          }
+        }
+    }
+  }
+
+  private def isPartialPartitionSpec(
+      partialSpec: TablePartitionSpec,
+      completedSpec: TablePartitionSpec): Boolean = {
+    partialSpec.forall {
+      case (partitionColumn, value) => completedSpec(partitionColumn) == value
```

Comment:
If the specified column names are not partitioning columns, we will get a strange error:
```
key not found: height
java.util.NoSuchElementException: key not found: height
```